pkg/operator/ceph/cluster/mon/spec.go

```diff
@@ -90,22 +104,212 @@ func (c *Cluster) makeDeployment(config *monConfig, hostname string) *extensions
+	}
+	monmapFilePath := path.Join(
+		mondaemon.GetMonRunDirPath(c.context.ConfigDir, config.DaemonName),
+		monmapFile,
+	)
+
+	monCommonArgs := []string{ // args needed for all ceph-mon calls
+		"--name", fmt.Sprintf("mon.%s", config.DaemonName),
+		"--cluster", c.clusterInfo.Name,
+		"--mon-data", mondaemon.GetMonDataDirPath(c.context.ConfigDir, config.DaemonName),
+		// It's safe to use the default path for the config b/c 'GenerateConfigFile'
+		// also writes a copy of the config to the default path
+		"--conf", cephconfig.DefaultConfigFilePath(),
+		// It's similarly safe to use the default path for the keyring b/c
+		// 'CreateKeyring' also writes a copy of the keyring to the default path
+		"--keyring", cephconfig.DefaultKeyringFilePath(),
+	}
+
+	// container := c.monContainer(config, c.clusterInfo.FSID)
 	podSpec := v1.PodSpec{
-		Containers:    []v1.Container{container},
+		InitContainers: []v1.Container{
+			{ // Config file init performed by Rook
+				Name: "config-init",
+				Args: []string{
+					"ceph",
+					"mon",
+					fmt.Sprintf("--config-dir=%s", k8sutil.DataDir),
+					fmt.Sprintf("--name=%s", config.DaemonName),
+					fmt.Sprintf("--port=%d", config.Port),
+					fmt.Sprintf("--fsid=%s", c.clusterInfo.FSID),
+				},
+				Image:           k8sutil.MakeRookImage(c.Version),
+				Env:             c.rookEnvVars(config),
+				VolumeMounts:    rookMounts(),
+				SecurityContext: podSecurityContext(),
+				// Init container doesn't need to expose ports since it does not communicate
+				Resources: c.resources,
+			},
+			{ // Ceph monmap init performed by Ceph 'monmaptool'
+				Name: "monmap-init",
+				Command: []string{
+					monmaptoolCommand,
+				},
+				Args: append(
+					[]string{
+						monmapFilePath,
+						"--create",
+						"--clobber",
+						"--fsid", c.clusterInfo.FSID,
+					},
+					monmapAddMonArgs...,
+				),
+				Image: k8sutil.MakeRookImage(c.Version), // TODO: ceph:<vers> image
+				// Containers which run daemons do not need env vars
+				VolumeMounts:    cephMounts(),
+				SecurityContext: podSecurityContext(),
+				Ports:           daemonPorts(config), // TODO: does monmaptool really need ports open?
+				Resources:       c.resources,
+			},
+			{ // Ceph mon filesystem init performed by mon daemon
+				Name: "mon-fs-init",
+				Command: []string{
+					cephMonCommand,
+				},
+				Args: append(
+					[]string{
+						"--mkfs",
+						"--monmap", monmapFilePath,
+					},
+					monCommonArgs...,
+				),
+				Image: k8sutil.MakeRookImage(c.Version), // TODO: ceph:<vers> image
+				// Containers which run daemons do not need env vars
+				VolumeMounts:    cephMounts(),
+				SecurityContext: podSecurityContext(),
+				Ports:           daemonPorts(config), // TODO: Does mkfs really need ports open?
+				Resources:       c.resources,
+			},
+		},
+		Containers: []v1.Container{
+			{
+				// The operator has set up the mon's service already, so we know the IP that the mon
```

Comment:
@travisn, this big comment block explains why I use the command `bash -c "ceph-mon ..."` here. I don't really like needing to use bash, but I don't see any alternatives. I looked around the Ceph config and k8s docs a lot today to see if I could do this without bash. Maybe you know something else I could try?

Reply:
Could we use the same pattern with `tini` as the osd uses on [this line](https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/osd/pod.go#L163)?

Reply:
I wish. `tini` isn't installed in the ceph-container containers.

Reply:
hmm, important detail that `tini` isn't there...

I believe that all of the ceph options can be specified either with a command line option, in the ceph.conf, or as an environment variable (although I'm not sure about this option). Can you try adding `public_bind_addr` and `public_addr` to the ceph.conf from the init container instead of adding the command line options to the main container? @liewegas these should work as config options, right? I see the variables are defined [here](https://github.com/ceph/ceph/blob/master/src/common/options.cc#L425-L429) in the ceph codebase.